Mathlib/Analysis/LocallyConvex/Barrelled.lean

```diff
@@ -0,0 +1,193 @@
+/-
+Copyright (c) 2023 Anatole Dedecker. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Anatole Dedecker
+-/
+import Mathlib.Analysis.LocallyConvex.WithSeminorms
+import Mathlib.Topology.Semicontinuous
+import Mathlib.Topology.MetricSpace.Baire
+
+/-!
+# Barrelled spaces and the Banach-Steinhaus theorem
+
+This files defines barrelled spaces over a `NontriviallyNormedField`, and proves the
+Banach-Steinhaus theorem for maps from a barrelled space to a space equipped with a family
+of seminorms generating the topology (i.e `WithSeminorms q` for some family of seminorms `q`).
+
+The more standard Banach-Steinhaus theorem for normed spaces is then deduced from that in
+`Analysis.NormedSpace.BanachSteinhaus`.
+
+## Main definitions
+
+* `BarrelledSpace`: a topological vector space `E` is said to be **barrelled** if all lower
+  semicontinuous seminorms on `E` are actually continuous. See the implementation details below for
+  more comments on this definition.
+* `WithSeminorms.continuousLinearMapOfTendsto`: fix `E` a barrelled space and `F` a TVS
+  satisfying `WithSeminorms q` for some `q`. Given a sequence of continuous linear maps from
+  `E` to `F` that converges pointwise to a function `f : E → F`, this bundles `f` as a
+  continuous linear map using the Banach-Steinhaus theorem.
+
+## Main theorems
+
+* `BaireSpace.instBarrelledSpace`: any TVS that is also a `BaireSpace` is barrelled. In
+  particular, this applies to Banach spaces and Fréchet spaces.
+* `WithSeminorms.banach_steinhaus`: fix `E` a barrelled space and `F` a TVS satisfying
+  `WithSeminorms q` for some `q`. Any family `𝓕 : ι → E →L[𝕜] F` of continuous linear maps
+  that is pointwise bounded is (uniformly) equicontinuous. Here, pointwise bounded means that
+  for all `k` and `x`, the family of real numbers `i ↦ q k (𝓕 i x)` is bounded above,
+  which is equivalent to requiring that `𝓕` is pointwise Von Neumann bounded
+  (see `WithSeminorms.image_isVonNBounded_iff_seminorm_bounded`).
+
+## Implementation details
+
+Barrelled spaces are defined in Bourbaki as locally convex spaces where barrels (aka closed
+balanced absorbing convex sets) are neighborhoods of zero. One can then show that barrels in a
+locally convex space are exactly closed unit balls of lower semicontinuous seminorms, hence that a
+locally convex space is barrelled iff any lower semicontinuous seminorm is continuous.
```

Comment:
Do you have plans to add the result  that `IsROrC 𝕜` implies `LocallyConvexSpace E ↔ BarrelledSpace 𝕜 E`?

Reply:
I don't think that's true. Does the docstring sound like it would be true?

The usual counterexample is to take the normed space of continuous functions on `[-1, 1]` equipped with the L2 norm (so it's definitely not complete), and consider the sequence of continuous linear forms defined by integrating against an approximate identity with L2 norm equal to one. That sequence of continuous linear maps converges pointwise to the evaluation at 0, yet evaluation at zero is not continuous in L2 norm, which contradicts the Banach Steinhaus theorem.

Reply:
I'm no analyst but to me the clause "a locally convex space is barrelled iff any lower semicontinuous seminorm is continuous" says what I wrote. Is the statement true assuming completeness?

Reply:
No worries!

I don't think completeness can help here without metrizability (in which case you have a Baire space so this PR covers it), but I don't know any particular counterexample.